File: packages/py-moose-lib/moose_lib/dmv2/stream.py
"""
Stream definitions for Moose Data Model v2 (dmv2).

This module provides classes for defining and configuring data streams,
including stream transformations, consumers, and dead letter queues.
"""
import dataclasses
import datetime
from typing import Any, Optional, Callable, Union, Literal, Generic
from pydantic import BaseModel, ConfigDict, AliasGenerator
from pydantic.alias_generators import to_camel

from .types import TypedMooseResource, ZeroOrMany, T, U
from .olap_table import OlapTable
from ._registry import _streams

class StreamConfig(BaseModel):
    """Configuration for data streams (e.g., Redpanda topics).

    Attributes:
        parallelism: Number of partitions for the stream.
        retention_period: Data retention period in seconds (default: 7 days).
        destination: Optional `OlapTable` where stream messages should be automatically ingested.
        version: Optional version string for tracking configuration changes.
        metadata: Optional metadata for the stream.
    """
    parallelism: int = 1
    retention_period: int = 60 * 60 * 24 * 7  # 7 days
    destination: Optional[OlapTable[Any]] = None
    version: Optional[str] = None
    metadata: Optional[dict] = None

class TransformConfig(BaseModel):
    """Configuration for stream transformations.

    Attributes:
        version: Optional version string to identify a specific transformation.
                 Allows multiple transformations to the same destination if versions differ.
    """
    version: Optional[str] = None
    dead_letter_queue: "Optional[DeadLetterQueue]" = None
    model_config = ConfigDict(arbitrary_types_allowed=True)
    metadata: Optional[dict] = None

class ConsumerConfig(BaseModel):
    """Configuration for stream consumers.

    Attributes:
        version: Optional version string to identify a specific consumer.
                 Allows multiple consumers if versions differ.
    """
    version: Optional[str] = None
    dead_letter_queue: "Optional[DeadLetterQueue]" = None
    model_config = ConfigDict(arbitrary_types_allowed=True)

@dataclasses.dataclass
class _RoutedMessage:
    """Internal class representing a message routed to a specific stream."""
    destination: "Stream[Any]"
    values: ZeroOrMany[Any]

@dataclasses.dataclass
class ConsumerEntry(Generic[T]):
    """Internal class representing a consumer with its configuration."""
    consumer: Callable[[T], None]
    config: ConsumerConfig

@dataclasses.dataclass
class TransformEntry(Generic[T]):
    """Internal class representing a transformation with its configuration."""
    destination: "Stream[Any]"
    transformation: Callable[[T], ZeroOrMany[Any]]
    config: TransformConfig

class Stream(TypedMooseResource, Generic[T]):
    """Represents a data stream (e.g., a Redpanda topic) typed with a Pydantic model.

    Allows defining transformations to other streams and adding consumers.

    Args:
        name: The name of the stream.
        config: Configuration options for the stream (parallelism, retention, destination).
        t: The Pydantic model defining the stream message schema (passed via `Stream[MyModel](...)`).

    Attributes:
        config (StreamConfig): Configuration settings for this stream.
        transformations (dict[str, list[TransformEntry[T]]]): Dictionary mapping destination stream names
                                                            to lists of transformation functions.
        consumers (list[ConsumerEntry[T]]): List of consumers attached to this stream.
        columns (Columns[T]): Helper for accessing message field names safely.
        name (str): The name of the stream.
        model_type (type[T]): The Pydantic model associated with this stream.
    """
    config: StreamConfig
    transformations: dict[str, list[TransformEntry[T]]]
    consumers: list[ConsumerEntry[T]]
    _multipleTransformations: Optional[Callable[[T], list[_RoutedMessage]]] = None

    def __init__(self, name: str, config: StreamConfig = StreamConfig(), **kwargs):
        super().__init__()
        self._set_type(name, self._get_type(kwargs))
        self.config = config
        self.metadata = config.metadata
        self.consumers = []
        self.transformations = {}
        _streams[name] = self

    def add_transform(self, destination: "Stream[U]", transformation: Callable[[T], ZeroOrMany[U]],
                      config: TransformConfig = None):
        """Adds a transformation step from this stream to a destination stream.

        The transformation function receives a record of type `T` and should return
        a record of type `U`, a list of `U` records, or `None` to filter.

        Args:
            destination: The target `Stream` for the transformed records.
            transformation: A callable that performs the transformation.
            config: Optional configuration, primarily for setting a version.
        """
        config = config or TransformConfig()
        if destination.name in self.transformations:
            existing_transforms = self.transformations[destination.name]
            # Check if a transform with this version already exists
            has_version = any(t.config.version == config.version for t in existing_transforms)
            if not has_version:
                existing_transforms.append(
                    TransformEntry(destination=destination, transformation=transformation, config=config))
        else:
            self.transformations[destination.name] = [
                TransformEntry(destination=destination, transformation=transformation, config=config)]

    def add_consumer(self, consumer: Callable[[T], None], config: ConsumerConfig = None):
        """Adds a consumer function to be executed for each record in the stream.

        Consumers are typically used for side effects like logging or triggering external actions.

        Args:
            consumer: A callable that accepts a record of type `T`.
            config: Optional configuration, primarily for setting a version.
        """
        config = config or ConsumerConfig()
        has_version = any(c.config.version == config.version for c in self.consumers)
        if not has_version:
            self.consumers.append(ConsumerEntry(consumer=consumer, config=config))

    def has_consumers(self) -> bool:
        """Checks if any consumers have been added to this stream.

        Returns:
            True if the stream has one or more consumers, False otherwise.
        """
        return len(self.consumers) > 0

    def routed(self, values: ZeroOrMany[T]) -> _RoutedMessage:
        """Creates a `_RoutedMessage` for use in multi-transform functions.

        Wraps the value(s) to be sent with this stream as the destination.

        Args:
            values: A single record, a list of records, or None.

        Returns:
            A `_RoutedMessage` object.
        """
        return _RoutedMessage(destination=self, values=values)

    def set_multi_transform(self, transformation: Callable[[T], list[_RoutedMessage]]):
        """Sets a transformation function capable of routing records to multiple streams.

        The provided function takes a single input record (`T`) and must return a list
        of `_RoutedMessage` objects, created using the `.routed()` method of the
        target streams.

        Example:
            def my_multi_transform(record: InputModel) -> list[_RoutedMessage]:
                output1 = transform_for_stream1(record)
                output2 = transform_for_stream2(record)
                return [
                    stream1.routed(output1),
                    stream2.routed(output2)
                ]
            input_stream.set_multi_transform(my_multi_transform)

        Note: Only one multi-transform function can be set per stream.

        Args:
            transformation: The multi-routing transformation function.
        """
        self._multipleTransformations = transformation

class DeadLetterModel(BaseModel, Generic[T]):
    """Model for dead letter queue messages.

    Attributes:
        original_record: The original record that failed processing.
        error_message: Description of the error that occurred.
        error_type: Type of error (e.g., "ValidationError").
        failed_at: Timestamp when the error occurred.
        source: Source of the error ("api", "transform", or "table").
    """
    model_config = ConfigDict(alias_generator=AliasGenerator(
        serialization_alias=to_camel,
    ))
    original_record: Any
    error_message: str
    error_type: str
    failed_at: datetime.datetime
    source: Literal["api", "transform", "table"]

    def as_typed(self) -> T:
        return self._t.model_validate(self.original_record)

class DeadLetterQueue(Stream, Generic[T]):
    """A specialized Stream for handling failed records.

    Dead letter queues store records that failed during processing, along with
    error information to help diagnose and potentially recover from failures.

    Attributes:
        All attributes inherited from Stream.
    """

    _model_type: type[T]

    def __init__(self, name: str, config: StreamConfig = StreamConfig(), **kwargs):
        """Initialize a new DeadLetterQueue.

        Args:
            name: The name of the dead letter queue stream.
            config: Configuration for the stream.
        """
        self._model_type = self._get_type(kwargs)
        kwargs["t"] = DeadLetterModel[self._model_type]
        super().__init__(name, config, **kwargs)

    def add_transform(self, destination: Stream[U], transformation: Callable[[DeadLetterModel[T]], ZeroOrMany[U]],
                      config: TransformConfig = None):
        def wrapped_transform(record: DeadLetterModel[T]):
            record._t = self._model_type
            return transformation(record)

        config = config or TransformConfig()
        super().add_transform(destination, wrapped_transform, config)

    def add_consumer(self, consumer: Callable[[DeadLetterModel[T]], None], config: ConsumerConfig = None):
        def wrapped_consumer(record: DeadLetterModel[T]):
            record._t = self._model_type
            return consumer(record)

        config = config or ConsumerConfig()
        super().add_consumer(wrapped_consumer, config)

    def set_multi_transform(self, transformation: Callable[[DeadLetterModel[T]], list[_RoutedMessage]]):
        def wrapped_transform(record: DeadLetterModel[T]):
            record._t = self._model_type
            return transformation(record)

        super().set_multi_transform(wrapped_transform)

File: packages/ts-moose-lib/src/dmv2/sdk/stream.ts
/**
 * @fileoverview Stream SDK for data streaming operations in Moose.
 *
 * This module provides the core streaming functionality including:
 * - Stream creation and configuration
 * - Message transformations between streams
 * - Consumer registration for message processing
 * - Dead letter queue handling for error recovery
 *
 * @module Stream
 */

import { IJsonSchemaCollection } from "typia";
import { TypedBase } from "../typedBase";
import { Column } from "../../dataModels/dataModelTypes";
import { dlqColumns, dlqSchema, getMooseInternal } from "../internal";
import { OlapTable } from "./olapTable";
import { LifeCycle } from "./lifeCycle";

/**
 * Represents zero, one, or many values of type T.
 * Used for flexible return types in transformations where a single input
 * can produce no output, one output, or multiple outputs.
 *
 * @template T The type of the value(s)
 * @example
 * ```typescript
 * // Can return a single value
 * const single: ZeroOrMany<string> = "hello";
 *
 * // Can return an array
 * const multiple: ZeroOrMany<string> = ["hello", "world"];
 *
 * // Can return null/undefined to filter out
 * const filtered: ZeroOrMany<string> = null;
 * ```
 */
export type ZeroOrMany<T> = T | T[] | undefined | null;

/**
 * Function type for transforming records from one type to another.
 * Supports both synchronous and asynchronous transformations.
 *
 * @template T The input record type
 * @template U The output record type
 * @param record The input record to transform
 * @returns The transformed record(s), or null/undefined to filter out
 *
 * @example
 * ```typescript
 * const transform: SyncOrAsyncTransform<InputType, OutputType> = (record) => {
 *   return { ...record, processed: true };
 * };
 * ```
 */
export type SyncOrAsyncTransform<T, U> = (
  record: T,
) => ZeroOrMany<U> | Promise<ZeroOrMany<U>>;

/**
 * Function type for consuming records without producing output.
 * Used for side effects like logging, external API calls, or database writes.
 *
 * @template T The record type to consume
 * @param record The record to process
 * @returns Promise<void> or void
 *
 * @example
 * ```typescript
 * const consumer: Consumer<UserEvent> = async (event) => {
 *   await sendToAnalytics(event);
 * };
 * ```
 */
export type Consumer<T> = (record: T) => Promise<void> | void;

/**
 * Configuration options for stream transformations.
 *
 * @template T The type of records being transformed
 */
export interface TransformConfig<T> {
  /**
   * Optional version identifier for this transformation.
   * Multiple transformations to the same destination can coexist with different versions.
   */
  version?: string;

  /**
   * Optional metadata for documentation and tracking purposes.
   */
  metadata?: { description?: string };

  /**
   * Optional dead letter queue for handling transformation failures.
   * Failed records will be sent to this queue for manual inspection or reprocessing.
   */
  deadLetterQueue?: DeadLetterQueue<T>;
}

/**
 * Configuration options for stream consumers.
 *
 * @template T The type of records being consumed
 */
export interface ConsumerConfig<T> {
  /**
   * Optional version identifier for this consumer.
   * Multiple consumers can coexist with different versions.
   */
  version?: string;

  /**
   * Optional dead letter queue for handling consumer failures.
   * Failed records will be sent to this queue for manual inspection or reprocessing.
   */
  deadLetterQueue?: DeadLetterQueue<T>;
}

/**
 * Represents a message routed to a specific destination stream.
 * Used internally by the multi-transform functionality to specify
 * where transformed messages should be sent.
 *
 * @internal
 */
class RoutedMessage {
  /** The destination stream for the message */
  destination: Stream<any>;

  /** The message value(s) to send */
  values: ZeroOrMany<any>;

  /**
   * Creates a new routed message.
   *
   * @param destination The target stream
   * @param values The message(s) to route
   */
  constructor(destination: Stream<any>, values: ZeroOrMany<any>) {
    this.destination = destination;
    this.values = values;
  }
}

/**
 * Configuration options for a data stream (e.g., a Redpanda topic).
 * @template T The data type of the messages in the stream.
 */
export interface StreamConfig<T> {
  /**
   * Specifies the number of partitions for the stream. Affects parallelism and throughput.
   */
  parallelism?: number;
  /**
   * Specifies the data retention period for the stream in seconds. Messages older than this may be deleted.
   */
  retentionPeriod?: number; // seconds
  /**
   * An optional destination OLAP table where messages from this stream should be automatically ingested.
   */
  destination?: OlapTable<T>;
  /**
   * An optional version string for this configuration. Can be used for tracking changes or managing deployments.
   */
  version?: string;
  metadata?: { description?: string };
  lifeCycle?: LifeCycle;
}

/**
 * Represents a data stream, typically corresponding to a Redpanda topic.
 * Provides a typed interface for producing to and consuming from the stream, and defining transformations.
 *
 * @template T The data type of the messages flowing through the stream. The structure of T defines the message schema.
 */
export class Stream<T> extends TypedBase<T, StreamConfig<T>> {
  /**
   * Creates a new Stream instance.
   * @param name The name of the stream. This name is used for the underlying Redpanda topic.
   * @param config Optional configuration for the stream.
   */
  constructor(name: string, config?: StreamConfig<T>);

  /** @internal **/
  constructor(
    name: string,
    config: StreamConfig<T>,
    schema: IJsonSchemaCollection.IV3_1,
    columns: Column[],
  );

  constructor(
    name: string,
    config?: StreamConfig<T>,
    schema?: IJsonSchemaCollection.IV3_1,
    columns?: Column[],
  ) {
    super(name, config ?? {}, schema, columns);

    const version = config?.version;
    const streamKey = version ? `${name}_${version}` : name;

    const streams = getMooseInternal().streams;
    if (streams.has(streamKey)) {
      throw new Error(`Stream with name ${streamKey} already exists`);
    }
    streams.set(streamKey, this);
  }

  /**
   * Internal map storing transformation configurations.
   * Maps destination stream names to arrays of transformation functions and their configs.
   *
   * @internal
   */
  _transformations = new Map<
    string,
    [Stream<any>, SyncOrAsyncTransform<T, any>, TransformConfig<T>][]
  >();

  /**
   * Internal function for multi-stream transformations.
   * Allows a single transformation to route messages to multiple destinations.
   *
   * @internal
   */
  _multipleTransformations?: (record: T) => [RoutedMessage];

  /**
   * Internal array storing consumer configurations.
   *
   * @internal
   */
  _consumers = new Array<{
    consumer: Consumer<T>;
    config: ConsumerConfig<T>;
  }>();

  /**
   * Adds a transformation step that processes messages from this stream and sends the results to a destination stream.
   * Multiple transformations to the same destination stream can be added if they have distinct `version` identifiers in their config.
   *
   * @template U The data type of the messages in the destination stream.
   * @param destination The destination stream for the transformed messages.
   * @param transformation A function that takes a message of type T and returns zero or more messages of type U (or a Promise thereof).
   *                       Return `null` or `undefined` or an empty array `[]` to filter out a message. Return an array to emit multiple messages.
   * @param config Optional configuration for this specific transformation step, like a version.
   */
  addTransform<U>(
    destination: Stream<U>,
    transformation: SyncOrAsyncTransform<T, U>,
    config?: TransformConfig<T>,
  ) {
    const transformConfig = config ?? {};

    if (this._transformations.has(destination.name)) {
      const existingTransforms = this._transformations.get(destination.name)!;
      const hasVersion = existingTransforms.some(
        ([_, __, cfg]) => cfg.version === transformConfig.version,
      );

      if (!hasVersion) {
        existingTransforms.push([destination, transformation, transformConfig]);
      }
    } else {
      this._transformations.set(destination.name, [
        [destination, transformation, transformConfig],
      ]);
    }
  }

  /**
   * Adds a consumer function that processes messages from this stream.
   * Multiple consumers can be added if they have distinct `version` identifiers in their config.
   *
   * @param consumer A function that takes a message of type T and performs an action (e.g., side effect, logging). Should return void or Promise<void>.
   * @param config Optional configuration for this specific consumer, like a version.
   */
  addConsumer(consumer: Consumer<T>, config?: ConsumerConfig<T>) {
    const consumerConfig = config ?? {};
    const hasVersion = this._consumers.some(
      (existing) => existing.config.version === consumerConfig.version,
    );

    if (!hasVersion) {
      this._consumers.push({ consumer, config: consumerConfig });
    }
  }

  /**
   * Helper method for `addMultiTransform` to specify the destination and values for a routed message.
   * @param values The value or values to send to this stream.
   * @returns A `RoutedMessage` object associating the values with this stream.
   *
   * @example
   * ```typescript
   * sourceStream.addMultiTransform((record) => [
   *   destinationStream1.routed(transformedRecord1),
   *   destinationStream2.routed([record2a, record2b])
   * ]);
   * ```
   */
  routed = (values: ZeroOrMany<T>) => new RoutedMessage(this, values);

  /**
   * Adds a single transformation function that can route messages to multiple destination streams.
   * This is an alternative to adding multiple individual `addTransform` calls.
   * Only one multi-transform function can be added per stream.
   *
   * @param transformation A function that takes a message of type T and returns an array of `RoutedMessage` objects,
   *                       each specifying a destination stream and the message(s) to send to it.
   */
  addMultiTransform(transformation: (record: T) => [RoutedMessage]) {
    this._multipleTransformations = transformation;
  }
}

/**
 * Base model for dead letter queue entries.
 * Contains the original failed record along with error information.
 */
export interface DeadLetterModel {
  /** The original record that failed processing */
  originalRecord: Record<string, any>;

  /** Human-readable error message describing the failure */
  errorMessage: string;

  /** Classification of the error type (e.g., "ValidationError", "TransformError") */
  errorType: string;

  /** Timestamp when the failure occurred */
  failedAt: Date;

  /** The source component where the failure occurred */
  source: "api" | "transform" | "table";
}

/**
 * Enhanced dead letter model with type recovery functionality.
 * Extends the base model with the ability to recover the original typed record.
 *
 * @template T The original record type before failure
 */
export interface DeadLetter<T> extends DeadLetterModel {
  /**
   * Recovers the original record as its typed form.
   * Useful for reprocessing failed records with proper type safety.
   *
   * @returns The original record cast to type T
   */
  asTyped: () => T;
}

/**
 * Internal function to attach type guard functionality to dead letter records.
 *
 * @internal
 * @template T The original record type
 * @param dl The dead letter model to enhance
 * @param typeGuard Function to validate and cast the original record
 */
function attachTypeGuard<T>(
  dl: DeadLetterModel,
  typeGuard: (input: any) => T,
): asserts dl is DeadLetter<T> {
  (dl as any).asTyped = () => typeGuard(dl.originalRecord);
}

/**
 * Specialized stream for handling failed records (dead letters).
 * Provides type-safe access to failed records for reprocessing or analysis.
 *
 * @template T The original record type that failed processing
 *
 * @example
 * ```typescript
 * const dlq = new DeadLetterQueue<UserEvent>("user-events-dlq");
 *
 * dlq.addConsumer(async (deadLetter) => {
 *   const originalEvent = deadLetter.asTyped();
 *   console.log(`Failed event: ${deadLetter.errorMessage}`);
 *   // Potentially reprocess or alert
 * });
 * ```
 */
export class DeadLetterQueue<T> extends Stream<DeadLetterModel> {
  /**
   * Creates a new DeadLetterQueue instance.
   * @param name The name of the dead letter queue stream
   * @param config Optional configuration for the stream. The metadata property is always present and includes stackTrace.
   */
  constructor(name: string, config?: StreamConfig<DeadLetterModel>);

  /** @internal **/
  constructor(
    name: string,
    config: StreamConfig<DeadLetterModel>,
    validate: (originalRecord: any) => T,
  );

  constructor(
    name: string,
    config?: StreamConfig<DeadLetterModel>,
    typeGuard?: (originalRecord: any) => T,
  ) {
    if (typeGuard === undefined) {
      throw new Error(
        "Supply the type param T so that the schema is inserted by the compiler plugin.",
      );
    }

    super(name, config ?? {}, dlqSchema, dlqColumns);
    this.typeGuard = typeGuard;
    getMooseInternal().streams.set(name, this);
  }

  /**
   * Internal type guard function for validating and casting original records.
   *
   * @internal
   */
  private typeGuard: (originalRecord: any) => T;

  /**
   * Adds a transformation step for dead letter records.
   * The transformation function receives a DeadLetter<T> with type recovery capabilities.
   *
   * @template U The output type for the transformation
   * @param destination The destination stream for transformed messages
   * @param transformation Function to transform dead letter records
   * @param config Optional transformation configuration
   */
  addTransform<U>(
    destination: Stream<U>,
    transformation: SyncOrAsyncTransform<DeadLetter<T>, U>,
    config?: TransformConfig<DeadLetterModel>,
  ) {
    const withValidate: SyncOrAsyncTransform<DeadLetterModel, U> = (
      deadLetter,
    ) => {
      attachTypeGuard<T>(deadLetter, this.typeGuard);
      return transformation(deadLetter);
    };
    super.addTransform(destination, withValidate, config);
  }

  /**
   * Adds a consumer for dead letter records.
   * The consumer function receives a DeadLetter<T> with type recovery capabilities.
   *
   * @param consumer Function to process dead letter records
   * @param config Optional consumer configuration
   */
  addConsumer(
    consumer: Consumer<DeadLetter<T>>,
    config?: ConsumerConfig<DeadLetterModel>,
  ) {
    const withValidate: Consumer<DeadLetterModel> = (deadLetter) => {
      attachTypeGuard<T>(deadLetter, this.typeGuard);
      return consumer(deadLetter);
    };
    super.addConsumer(withValidate, config);
  }

  /**
   * Adds a multi-stream transformation for dead letter records.
   * The transformation function receives a DeadLetter<T> with type recovery capabilities.
   *
   * @param transformation Function to route dead letter records to multiple destinations
   */
  addMultiTransform(
    transformation: (record: DeadLetter<T>) => [RoutedMessage],
  ) {
    const withValidate: (record: DeadLetterModel) => [RoutedMessage] = (
      deadLetter,
    ) => {
      attachTypeGuard<T>(deadLetter, this.typeGuard);
      return transformation(deadLetter);
    };
    super.addMultiTransform(withValidate);
  }
}
